content/src/content/jcr_root/apps/core/wcm/components/commons/datalayer/clientlibs/js/datalayer.js

```diff
@@ -0,0 +1,392 @@
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+(function() {
+    "use strict";
+
+    /* eslint no-console: "off" */
+    /* eslint no-unused-vars: "off" */
+
+    /**
+     * @typedef {String} DataLayerEvents
+     **/
+
+    /**
+     * Enumeration of data layer events.
+     *
+     * @enum {DataLayerEvents}
+     * @readonly
+     */
+    var events = {
+        /** Represents an event triggered for any change in the data layer state */
+        CHANGE: "datalayer:change",
+        /** Represents an event triggered for any event push to the data layer */
+        EVENT: "datalayer:event",
+        /** Represents an event triggered when the data layer has initialized */
+        READY: "datalayer:ready"
+    };
+
+    /**
+     * @typedef {String} ListenerScope
+     **/
+
+    /**
+     * Enumeration of listener scopes.
+     *
+     * @enum {ListenerScope}
+     * @readonly
+     */
+    var listenerScope = {
+        /** Past events only */
+        PAST: "past",
+        /** Future events only */
+        FUTURE: "future",
+        /** All events, past and future */
+        ALL: "all"
+    };
+
+    /**
+     * @typedef  {Object} ListenerOnConfig
+     * @property {String} on Name of the event to bind to.
+     * @property {String} [selector] Object key in the state to bind to.
+     * @property {ListenerScope} [scope] Scope of the listener.
+     * @property {Function} handler Handler to execute when the bound event is triggered.
+     */
+
+    /**
+     * @typedef  {Object} ListenerOffConfig
+     * @property {String} off Name of the event to unbind.
+     * @property {Function} [handler] Handler for a previously attached event to unbind.
+     */
+
+    /**
+     * @typedef {Object} DataConfig
+     * @property {Object} data Data to be updated in the state.
+     */
+
+    /**
+     * @typedef {Object} EventConfig
+     * @property {String} name Name of the event.
+     * @property {Object} [info] Additional information to pass to the event handler.
+     * @property {DataConfig.data} [data] Data to be updated in the state.
+     */
+
+    /**
+     * @typedef {DataConfig | EventConfig | ListenerOnConfig | ListenerOffConfig} ItemConfig
+     */
+
+    /**
+     * @class DataLayer
+     * @classdesc Data layer controller that augments the passed data layer array and handles eventing.
+     * @param {Array} dataLayer The data layer array.
+     */
+    function DataLayer(dataLayer) {
+        var that = this;
+
+        that.dataLayer = dataLayer;
+        that.state = {};
+
+        /**
+         * Returns a deep copy of the data layer state.
+         *
+         * @returns {Object} The deep copied state object.
+         */
+        that.dataLayer.getState = function() {
+            // use deep copying technique of JSON stringify and parsing the state.
+            return JSON.parse(JSON.stringify(that.state));
```

Comment:
intention is to clone state, so that if you alternate returned value will now change original object?
asking, because stringify + parse is pretty slow

Reply:
https://github.com/immutable-js/immutable-js might be an alternative to produce immutable objects. My Javascript fu is not sufficient to judge how good that is, I'll leave that to specialists!